GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+        private int bufferContentLength;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        public string ReadMessage()
+        {
+            List<byte> bytes = new List<byte>();
+
+            int bytesRead = this.Read();
+            if (bytesRead == 0)
+            {
+                return null;
+            }
+
+            while (true)
+            {
+                if (this.buffer[bytesRead - 1] == TerminatorByte)
+                {
+                    bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead - 1));
+                    break;
+                }
+
+                bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead));
+                bytesRead = this.Read();
+                Debug.Assert(bytesRead != 0, "Not expecting 0 bytes at this time");
+            }
+
+            return Encoding.UTF8.GetString(bytes.ToArray());

```

Comment:
It seems like the real value add of this overall change is using a different terminator character. What additional protection do we get from using UTF8 encoding? At first glance, it appears that this will just add perf overhead.

Reply:
Good points - you are correct - the main value of this change is using a different terminator character. We don't need to use UTF8 (it doesn't add any additional protection) - we just need the sender / receiver to agree on the encoding. There are some reasons I chose UTF8 for this iteration:

1. GVFS currently sends bytes as UTF8 (implicitly via the `StreamReader` / `StreamWriter` classes). Using UTF8 again would be keeping with current behavior, so there shouldn't be any additional overhead.

2. The native code that reads / writes data from the named pipe treats the data as ASCII (which I understand is a subset of UTF8). If we switched away from UTF8, we would need to do convert the encoding in the native code.

3. The ModifiedPathsList hook also reads data from the named pipe and directly sends those bytes to git, which works with UTF8 (at least for our current scenarios), but would not work with UTF16 (which is the default representation of strings in C#).

I propose we continue using UTF8 (explicitly now) for this change? We can consider if a different encoding would make more sense...